CIP-0139/README.md

```diff
@@ -0,0 +1,152 @@
+
+The only way to implement a builtin map with a (builtin) polymorphic key type would be to modify the builtins language itself. A compiler would have to only allow the construction of maps with key types which are known a priori to possess some property (such as a total ordering). The Plutus Core evaluator would have to check for this as well. The implementation itself would not be too complicated, but we must first consider how this would change the formal semantics of the builtins language.
+
+Given the above, the only immediate solution is to implement a builtin map type which is specialised to `ByteString` keys but remains polymorphic in the value type. We believe that this solution should be good enough for most practical use-cases, and maybe even surpass a fully polymorphic map type in performance.
+
+
+## Specification
+
+We propose the addition of the following builtin type: `ByteStringMap` of kind `Type -> Type`.
+- It represents a map associating keys of type `ByteString` with values of any builtin type.
+- The new builtin type should be implemented using a tree data structure allowing for logarithmic-time map operations.
+- We consider the costing size of a `ByteStringMap` to be equal to the size of its corresponding representation as an association list.
+- The binary encoding of `ByteStringMap`s is equivalent to its encoding as an association list.
+
+**Note**: Here `Type` is the universe of all builtin types, since we do not consider types formed out of applying builtin types to arbitrary types to be inhabited.
+
+We propose the following set of builtin functions to accompany the new builtin type:
+1. `insertBSMap :: forall v . ByteString -> v -> ByteStringMap v -> ByteStringMap v`
+    - it returns a map with the input key updated to the new value, silently discarding any previous value
+    - it uses logarithmic time in the `ByteStringMap` argument, linear time in the `ByteString` argument, and logarithmic memory in the `ByteStringMap` argument
+2. `lookupBSMap :: forall v . ByteString -> (() -> v) -> ByteStringMap v -> v`
+    - it returns the value for a given key in the map; if the key is missing, it returns the default value of type `() -> a` 
+    - its implementation depends on, as of the time of writing, ongoing work in supporting higher-order builtin functions; see the next section for the motivation behind this choice
+    - it uses logarithmic time in the `ByteStringMap` argument, linear time in the `ByteString` argument and constant memory
+3. `deleteBSMap :: forall v . ByteString -> ByteStringMap v -> ByteStringMap v`
+    - it returns a map without the key-value association identified by the input key
+    - it uses logarithmic time in the `ByteStringMap` argument, linear time in the `ByteString` argument and logarithmic memory in the `ByteStringMap` argument
+4. `unionBSMap :: forall v . ByteStringMap v -> ByteStringMap v -> ByteStringMap v`
```

Comment:
I wonder if we want `intersectionBSMap` and `differenceBSMap` then.

A horrible nit btw: I'd personally prefer a different capitalization: `unionBsMap`. But feel free to ignore.